pkg/httpd/webfolder.go

```diff
@@ -20,56 +20,67 @@ func (h *webFolder) Name() string {
 }
 
 func (h *webFolder) CheckValidation() error {
-	apiClient := h.ws.apiClient
-	user := h.ws.CurrentUser()
-	terminalCfg, err := h.ws.apiClient.GetTerminalConfig()
+	if volume, err := SftpCheckValidation(h.ws); err != nil {
+		return err
+	} else {
+		h.volume = volume
+	}
+	return nil
+}
+
+func (h *webFolder) HandleMessage(*Message) {
+}
+
+func (h *webFolder) CleanUp() {
+	close(h.done)
+	h.volume.Close()
+}
+
+func (h *webFolder) GetVolume() *UserVolume {
+	select {
+	case <-h.done:
+		return nil
+	default:
+		return h.volume
+	}
+}
+
+func SftpCheckValidation(ws *UserWebsocket) (*UserVolume, error) {
+	apiClient := ws.apiClient
+	user := ws.CurrentUser()
+	terminalCfg, err := ws.apiClient.GetTerminalConfig()
+
+	uv := &UserVolume{}
 	if err != nil {
 		logger.Errorf("Get terminal config failed: %s", err)
-		return err
+		return uv, err
 	}
 	volOpts := make([]VolumeOption, 0, 5)
 	volOpts = append(volOpts, WithUser(user))
-	volOpts = append(volOpts, WithAddr(h.ws.ClientIP()))
+	volOpts = append(volOpts, WithAddr(ws.ClientIP()))
 	volOpts = append(volOpts, WithTerminalCfg(&terminalCfg))
-	params := h.ws.wsParams
+	params := ws.wsParams
 	targetId := params.TargetId
 	assetId := params.AssetId
 	if assetId == "" {
 		assetId = targetId
 	}
-	if h.ws.ConnectToken != nil {
-		connectToken := h.ws.ConnectToken
+	if ws.ConnectToken != nil {
+		connectToken := ws.ConnectToken
 		volOpts = append(volOpts, WithConnectToken(connectToken))
 	} else {
 		if common.ValidUUIDString(assetId) {
 			assets, err := apiClient.GetUserAssetByID(user.ID, assetId)
 			if err != nil {
 				logger.Errorf("Get user asset %s error: %s", assetId, err)
-				return ErrAssetIdInvalid
+				return uv, ErrAssetIdInvalid
 			}
 			if len(assets) != 1 {
 				logger.Errorf("Get user more than one asset %s: choose first", targetId)
 			}
 			volOpts = append(volOpts, WithAsset(&assets[0]))
 		}
 	}
-	h.volume = NewUserVolume(apiClient, volOpts...)
-	return nil
-}
 
-func (h *webFolder) HandleMessage(*Message) {
-}
-
-func (h *webFolder) CleanUp() {
-	close(h.done)
-	h.volume.Close()
-}
-
-func (h *webFolder) GetVolume() *UserVolume {
-	select {
-	case <-h.done:
-		return nil
-	default:
-		return h.volume
-	}
+	return NewUserVolume(apiClient, volOpts...), nil
 }
```

Comment:
The following code has several inconsistencies and potential errors. Here are some of them:

1. The function `HandleMessage` should be defined elsewhere because it is not clear where this function needs to be used.

2. In the initialization block inside `handle`, there are two calls to "volume". This could lead to ambiguity since it's unclear whether volume refers to either the UserVolume object created earlier or another separate volume instance that was previously initialized.

3. There seems to be an issue with the handling of error conditions in the validation process for web folders. Error messages suggest that this can cause problems like getting a specific folder type, such as webfolder. 

4. There seems to be an incorrect way to create multiple instances of `NewUserVolume`. It appears you attempt to use this method recursively on a single volume option chain but don't ensure they're all created correctly.

To address these concerns:
   1. Define the interface `IMessageHandler`.
   2. Ensure the correct scope of variables across functions.
   3. Add proper validation checks when initializing volumes if possible. For example, check if provided user_id matches the one in config before attempting to access an asset ID, or similarly. 

Note: I've only listed these points here and haven't made any implementation changes directly above; your comments suggest that fixing these issues through design decisions would also require refactoring the structure of your existing codebase to align better with Go best practices.

For additional guidance, please refer closely to your project specifications and/or documentation. If necessary, discuss with team members who understand your particular context. Don't hesitate to consult official programming guidelines from your organization if they exist.